File: sqlparser/src/sql/TableFinder.java
//@
package cdata.sql;
import rssbus.RSBException;
import rssbus.oputils.common.Utilities;
//@
/*#
using RSSBus.core;
using RSSBus;
namespace CData.Sql {
#*/

final class TableFinder implements ISqlQueryVisitor {
  private final SqlCollection<SqlTable> _collector;
  private final ITableMatch _matcher;
  public TableFinder(SqlCollection<SqlTable> collector, ITableMatch matcher) {
    this._collector = collector;
    this._matcher = matcher;
  }

  public void visit(SqlTable table) throws RSBException {
    if (table.isNestedQueryTable()) {
      if (this._matcher.accept(table, TablePartType.NestedQuery)) {
        this._collector.add(this._matcher.create(table));
      }

      if (this._matcher.unwind(table, TablePartType.NestedQuery)) {
        SqlQueryStatement nq = table.getQuery();
        SqlCollection<SqlTable> NQ_REFERENCES = new SqlCollection<SqlTable>();
        TableFinder NQ_COLLECTOR = new TableFinder(NQ_REFERENCES, this._matcher);
        try {
          nq.accept(NQ_COLLECTOR);
        } catch (Exception ex) {
          throw new RSBException("TableCollector", Utilities.getExceptionMessage(ex));
        }
        this._collector.addAll(NQ_REFERENCES);
      }
    } else if (table.isNestedJoinTable()) {
      if (this._matcher.accept(table, TablePartType.NestedJoin)) {
        this._collector.add(this._matcher.create(table));
      }

      if (this._matcher.unwind(table, TablePartType.NestedJoin)) {
        SqlTable nj = table.getNestedJoin();
        SqlCollection<SqlTable> NJ_REFERENCES = new SqlCollection<SqlTable>();
        TableFinder NJ_COLLECTOR = new TableFinder(NJ_REFERENCES, this._matcher);
        try {
          nj.accept(NJ_COLLECTOR);
        } catch (Exception ex) {
          throw new RSBException("TableCollector", Utilities.getExceptionMessage(ex));
        }
        this._collector.addAll(NJ_REFERENCES);
      }
    } else {
      if (this._matcher.accept(table, TablePartType.SimpleTable)) {
        this._collector.add(this._matcher.create(table));
      }
      // THIS IS A HACK so that SqlQueryAliasChecker resolves
      // table aliases in the query correctly to columns that are tied
      // to our cross apply statements.
      SqlCrossApply ca = table.getCrossApply();
      while ( ca != null ) {
        SqlTable pseudoTable = ca.getPseudoTable();
        if (this._matcher.accept(pseudoTable, TablePartType.SimpleTable)) {
          this._collector.add(this._matcher.create(pseudoTable));
        }
        ca = ca.getCrossApply();
      }
    }

    if (table.hasJoin()) {
      SqlJoin j = table.getJoin();
      SqlTable right = j.getTable();
      this.visit(right);
    }
  }

  public void visit(SqlStatement stmt) throws RSBException {
    //do nothing.
  }

  public void visit(SqlColumn column) throws RSBException {
    //do nothing.
  }

  public void visit(SqlConditionNode criteria) throws RSBException {
    if (criteria instanceof SqlConditionExists) {
      SqlQueryStatement sq = ((SqlConditionExists) criteria).getSubQuery();
      try {
        sq.accept(this);
      } catch (Exception ex) {
        throw new RSBException("TableCollector", Utilities.getExceptionMessage(ex));
      }
    } else if (criteria instanceof SqlConditionInSelect) {
      SqlQueryStatement sq = ((SqlConditionInSelect) criteria).getRightQuery();
      try {
        sq.accept(this);
      } catch (Exception ex) {
        throw new RSBException("TableCollector", Utilities.getExceptionMessage(ex));
      }
    }
  }
}

File: sqlparser/src/sql/csgen/TableFinder.cs
using System;
using System.Collections;
using System.IO;
using System.Threading;
using System.Net;
using System.Net.Sockets;
using System.Text;
using RSSBus.core.j2cs;



using RSSBus.core;
using RSSBus;
namespace CData.Sql {


sealed class TableFinder : ISqlQueryVisitor {
  private readonly SqlCollection<SqlTable> _collector;
  private readonly ITableMatch _matcher;
  public TableFinder(SqlCollection<SqlTable> collector, ITableMatch matcher) {
    this._collector = collector;
    this._matcher = matcher;
  }

  public void Visit(SqlTable table) {
    if (table.IsNestedQueryTable()) {
      if (this._matcher.Accept(table, TablePartType.NestedQuery)) {
        this._collector.Add(this._matcher.Create(table));
      }

      if (this._matcher.Unwind(table, TablePartType.NestedQuery)) {
        SqlQueryStatement nq = table.GetQuery();
        SqlCollection<SqlTable> NQ_REFERENCES = new SqlCollection<SqlTable>();
        TableFinder NQ_COLLECTOR = new TableFinder(NQ_REFERENCES, this._matcher);
        try {
          nq.Accept(NQ_COLLECTOR);
        } catch (Exception ex) {
          throw new RSBException("TableCollector", Utilities.GetExceptionMessage(ex));
        }
        this._collector.AddAll(NQ_REFERENCES);
      }
    } else if (table.IsNestedJoinTable()) {
      if (this._matcher.Accept(table, TablePartType.NestedJoin)) {
        this._collector.Add(this._matcher.Create(table));
      }

      if (this._matcher.Unwind(table, TablePartType.NestedJoin)) {
        SqlTable nj = table.GetNestedJoin();
        SqlCollection<SqlTable> NJ_REFERENCES = new SqlCollection<SqlTable>();
        TableFinder NJ_COLLECTOR = new TableFinder(NJ_REFERENCES, this._matcher);
        try {
          nj.Accept(NJ_COLLECTOR);
        } catch (Exception ex) {
          throw new RSBException("TableCollector", Utilities.GetExceptionMessage(ex));
        }
        this._collector.AddAll(NJ_REFERENCES);
      }
    } else {
      if (this._matcher.Accept(table, TablePartType.SimpleTable)) {
        this._collector.Add(this._matcher.Create(table));
      }
      // THIS IS A HACK so that SqlQueryAliasChecker resolves
      // table aliases in the query correctly to columns that are tied
      // to our cross apply statements.
      SqlCrossApply ca = table.GetCrossApply();
      while ( ca != null ) {
        SqlTable pseudoTable = ca.GetPseudoTable();
        if (this._matcher.Accept(pseudoTable, TablePartType.SimpleTable)) {
          this._collector.Add(this._matcher.Create(pseudoTable));
        }
        ca = ca.GetCrossApply();
      }
    }

    if (table.HasJoin()) {
      SqlJoin j = table.GetJoin();
      SqlTable right = j.GetTable();
      this.Visit(right);
    }
  }

  public void Visit(SqlStatement stmt) {
    //do nothing.
  }

  public void Visit(SqlColumn column) {
    //do nothing.
  }

  public void Visit(SqlConditionNode criteria) {
    if (criteria is SqlConditionExists) {
      SqlQueryStatement sq = ((SqlConditionExists) criteria).GetSubQuery();
      try {
        sq.Accept(this);
      } catch (Exception ex) {
        throw new RSBException("TableCollector", Utilities.GetExceptionMessage(ex));
      }
    } else if (criteria is SqlConditionInSelect) {
      SqlQueryStatement sq = ((SqlConditionInSelect) criteria).GetRightQuery();
      try {
        sq.Accept(this);
      } catch (Exception ex) {
        throw new RSBException("TableCollector", Utilities.GetExceptionMessage(ex));
      }
    }
  }
}
}

